java/meterpreter/meterpreter/src/main/java/com/metasploit/meterpreter/Transport.java

```diff
@@ -45,6 +47,60 @@ protected int parseTimeouts(byte[] configuration, int offset) {
         return offset;
     }
 
+    protected TLVPacket readAndDecodePacket(DataInputStream in) throws IOException {
+        int xorKey = in.readInt();
+        int len = (in.readInt() ^ Integer.reverseBytes(xorKey)) - 8;
+        int type = in.readInt();
+        byte[] body = new byte[len];
+        in.readFully(body);
+
+        this.xorBytes(xorKey, body);
+
+        ByteArrayInputStream byteStream = new ByteArrayInputStream(body);
+        DataInputStream inputStream = new DataInputStream(byteStream);
+        TLVPacket packet = new TLVPacket(inputStream, len);
+        inputStream.close();
+
+        return packet;
+    }
+
+    protected void encodePacketAndWrite(TLVPacket packet, int type, DataOutputStream out) throws IOException {
+        int xorKey = randXorKey();
+        byte[] data = packet.toByteArray();
+        this.xorBytes(xorKey, data);
+        synchronized (out) {
+            out.writeInt(xorKey);
+            out.writeInt((data.length + 8) ^ Integer.reverseBytes(xorKey));
```

Comment:
Hmm, are we Java 1.5+ now? I'm surprised that Maven Animal Sniffer did not catch that. (Integer.reverseBytes is since 1.5, in 1.4 there was a similar method added in the NIO buffers library, but for older ones you are swapping the bytes yourself...).

On the other hand, I guess Java 1.4 is ancient enough that if you find it in a pentest, you find other easier targets too :)